### DISCUSSION TITLE
LSP in containers

### DISCUSSION THREAD

**Discussion by abitrolly** [w0rp/ale#3635] [Ideas]
I like to run every bit of potentially insecure code in containers and it would be pretty awesome if ALE could use containers to run LSP servers related to current project.

**Comment by w0rp** [comment]
`g:ale_filename_mappings` can probably be used to run commands in Docker containers provided files are mapped the right ways.

**Reply by abitrolly** [reply]
I am not familiar with Language Server Protocol, but isn't it the client who sends file contents to it?

Do you mean that LSP server needs a direct access to all files on my system?

**Reply by hsanson** [reply]
Cannot tell for all LSP servers but the ones I do use require access to the whole project in order to handle auto-completion, dependencies, and errors properly.

This makes sense since LSP servers wouldn't be able to determine if a class/function/variable is defined somewhere else in the project by looking at a single file in isolation.

**Reply by abitrolly** [reply]
True. Does the protocol support sending files at all? And along that, sharing the whole project dir?

**Reply by w0rp** [reply]
The LSP server ought not to need access to the filesystem, but _some servers_ might do. The LSP protocol lets you define servers that even run on remote TCP connections, which could be all the way around the world. I think most servers assume they are running on the same machine as the client.

**Reply by abitrolly** [reply]
I found that LSP supports sending files - https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocumentItem

What I didn't find is specification how an LSP should access filesystem directly. Or request other necessary files from the client. It looks like the client should be aware of language conventions and push artifacts like `Gemfile` proactively. Maybe I've missed something. There are some Workspace objects in the spec. Maybe client send a list of files to choose from. I do not see this workflow or user story described.
